share/completions/dnf.fish

```diff
@@ -1,5 +1,182 @@
 #
-# Completions for the dnf command
+# Completions for the dnf Fedora 27 
+# 
+# by celso.lom - 2018
 #
+# Need the sqlite3 package
 
-complete -c dnf -w yum
+# Check commands to return package list
+function __fish_dnf_package_installed
+	not __fish_seen_subcommand_from group; and __fish_seen_subcommand_from update upgrade remove erase reinstall
+end
+
+function __fish_dnf_package_notinstalled
+	not __fish_seen_subcommand_from group; and __fish_seen_subcommand_from install
+end
+
+function __fish_dnf_package_allavailable
+	not __fish_seen_subcommand_from group history; and __fish_seen_subcommand_from info deplist
+end
+
+
+# Return installed packages
+function __dnf_return_installed
+	sqlite3 /var/cache/dnf/packages.db "select pkg from installed WHERE pkg LIKE \"$cur%\"" 2>/dev/null
+end
+
+# Return available but not installed packages
+function __dnf_return_notinstalled
+	sqlite3 /var/cache/dnf/packages.db "select available.pkg from available left join installed on available.pkg = installed.pkg where installed.pkg IS NULL and available.pkg LIKE \"$cur%\"" 2>/dev/null
+end
+
+# Return all packages
+function __dnf_return_allavailable
+	sqlite3 /var/cache/dnf/packages.db "select pkg from available WHERE pkg LIKE \"$cur%\"" 2>/dev/null
+end
+
+
+# Return Groups
+#TODO: Completion of groups to slow, test other options
+# Available
+#for i in (dnf group list --available | grep -v ':' | sed -e "s/^ *//g" | )
+#	set -g dnfgroup_available "$dnfgroup_available\n$i"
+#end
+#
+## Installed
+#for i in (dnf group list --installed | grep -v ':' | sed -e "s/^ *//g")
+#	set -g dnfgroup_installed "$dnfgroup_installed\n$i"
+#end
+#
+## All
+#for i in (dnf group list | grep -v ':' | sed -e "s/^ *//g")
+#	set -g dnfgroup_all "$dnfgroup_all\n$i"
+#end
+#
+#function __dnf_return_groups
+#	switch $argv
+#		case "available"
+#			echo -e $dnfgroup_available
+#		case "installed"
+#			echo -e $dnfgroup_installed
+#		case "all"
+#			echo -e $dnfgroup_all
+#	end
+#end
+
+
+# Return help
+set -e dnfhelp
+for i in (dnf help)
+	set dnfhelp "$dnfhelp\n$i"
```

Comment:
I wrote a whole bunch of comments about this part, because it is weird (you're constructing a one-element variable and then later splitting it again later).

But the most important thing is that it is too clever. It's trying to parse the help and hoping that that is suitable for use as a description. It's also hoping that the description for a single subcommand is never more than a line.

Instead, I suggest just simply taking that output and using it to write descriptions, and put them into this file directly.